### PR TITLE
Format more text literals as multi-line strings

### PR DESCRIPTION
**Pull request by sjakobi**
This causes text literals to be formatted as multi-line strings whenever they contain at least one newline and at least one non-newline character. "Spacers" like `"\n\n"` continue be formatted as single-line strings. If the heuristic turns out to be too eager to choose a multi-line layout, we can refine it later.

This partially addresses https://github.com/dhall-lang/dhall-haskell/issues/1496.

### CODE REVIEW

**Comment by sjakobi** [comment]
Two tests currently trigger a `prettyprinter` bug (https://github.com/quchen/prettyprinter/issues/91). We could work around that by changing `pretty_` to use our usual 80-columns layout:

https://github.com/dhall-lang/dhall-haskell/blob/cc1814be9858f62e664b0d2114e711fd4431bd2c/dhall/src/Dhall/Pretty/Internal.hs#L1113-L1117

We also use `Unbounded` layouts in `prettyToString`, `docToStrictText`, and the diff tests:

https://github.com/dhall-lang/dhall-haskell/blob/cc1814be9858f62e664b0d2114e711fd4431bd2c/dhall/tests/Dhall/Test/Diff.hs#L53-L59

**Comment by Gabriel439** [comment]
@sjakobi: As discussed in the other issue, another way you can fix this is to ensure that every use of `Line`/`hardline` has a newline-free alternative provided via a `flatAlt`

**Comment by sjakobi** [comment]
> another way you can fix this is to ensure that every use of `Line`/`hardline` has a newline-free alternative provided via a `flatAlt`

It seems to me that we couldn't then enforce the multi-line formatting of the text literals that this PR is about.

If it is important that `pretty_` introduces as few line breaks as possible, we could instead use something like `AvailablePerLine maxBound 1.0` for the `LayoutOptions`.

These are the problematic tests:


#### 1

https://github.com/dhall-lang/dhall-haskell/blob/cc1814be9858f62e664b0d2114e711fd4431bd2c/dhall/tests/Dhall/Test/Regression.hs#L90-L93

Here we use `pretty_` via `Core.pretty` via `normalize'`:

https://github.com/dhall-lang/dhall-haskell/blob/cc1814be9858f62e664b0d2114e711fd4431bd2c/dhall/tests/Dhall/Test/Util.hs#L63-L64

#### 2

https://github.com/dhall-lang/dhall-haskell/blob/cc1814be9858f62e664b0d2114e711fd4431bd2c/dhall/tests/Dhall/Test/Tutorial.hs#L49-L57

We again use `normalize'` via `assertNormalized`.

---

I think my preference would be to simply use the 80-column layout consistently, so we don't have to worry about different layout variants.

**Comment by Gabriel439** [comment]
@sjakobi: I agree that the best solution here is to consistently avoid the `Unbounded` layout option, because it appears that it is incompatible with output that has obligate newlines

File: dhall/src/Dhall/Pretty/Internal.hs [hunk 1]
```diff
@@ -1164,14 +1164,16 @@ escapeText_ text = Text.concatMap adapt text
 
 prettyToString :: Pretty a => a -> String
 prettyToString =
-    Pretty.renderString . Pretty.layoutPretty options . Pretty.pretty
-  where
-   options = Pretty.LayoutOptions { Pretty.layoutPageWidth = Pretty.Unbounded }
+    Pretty.renderString . Pretty.layoutPretty layoutOpts . Pretty.pretty
```
Comment (sjakobi) [inline dhall/src/Dhall/Pretty/Internal.hs]: Uh, should we maybe also use `layoutSmart` consistently?!
```suggestion
    Pretty.renderString . Pretty.layoutSmart layoutOpts . Pretty.pretty
```

Comment (Gabriel439) [reply dhall/src/Dhall/Pretty/Internal.hs]: Yeah, that's fine

**Pull request merged**